File: python/ch18_agent/agent_ch18/features/work_stealing.py
"""第十八章任务认领 Service、租约工具和空闲轮询配置。

Java 分层对照：

* ``LeasedTaskStore`` 是 Repository interface；SQLite 只是它的一个 Adapter。
* ``DirectTaskClaimService`` 是领域 Service，把可信 identity 绑定成任务 owner。
* ``WorkStealingRuntime`` 是 WorkerService 的配置对象，统一保存轮询策略和工具快照。

模型永远不能在工具参数里填写 owner。owner 来自 ``ToolContext.identity``，而完成任务
还必须回传当前 ``claim_token``。这和“用户名 + 一次性乐观锁令牌”同时校验很相似。
"""

from __future__ import annotations

import json
import threading
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, Protocol

from ..core.tools import (
    ToolContext,
    ToolDefinition,
    ToolRegistry,
    ToolResult,
    tool_error,
    tool_success,
)
from .tasks import (
    CANONICAL_UUID,
    CreateTaskInput,
    Task,
    TaskCompletion,
    TaskError,
    TaskStorageError,
    canonical_task_id,
)


class TaskClaimError(TaskError):
    """owner 或 claim token 与当前有效认领不匹配。"""

    def __init__(self, message: str, code: str = "task_claim_mismatch") -> None:
        super().__init__(code, message)


class TaskLeaseExpiredError(TaskClaimError):
    """认领曾经有效，但租约已经到达半开区间的结束边界。"""

    def __init__(self, message: str) -> None:
        super().__init__(message, "task_lease_expired")


@dataclass(frozen=True, slots=True)
class TaskClaim:
    """一次成功认领的不可变结果。

    字段说明：
        ``task``：已经进入 in_progress 的任务快照。
        ``claim_token``：完成权限证明；旧 token 不能完成重新认领后的任务。
        ``lease_expires_at_utc``：本次认领的 UTC 到期时间。

    Java 可以把它理解成 ``record TaskClaim(Task task, UUID token, Instant expiresAt)``。
    """

    task: Task
    claim_token: str
    lease_expires_at_utc: datetime


class LeasedTaskStore(Protocol):
    """带租约的任务仓库接口，类似 Java Repository interface。"""

    def create_task(self, value: CreateTaskInput) -> Task:
        """创建一个带显式依赖的持久任务。"""

    def get_task(self, task_id: str) -> Task:
        """读取单个任务。"""

    def list_tasks(self) -> tuple[Task, ...]:
        """按稳定创建顺序列出任务图。"""

    def claim_task(self, task_id: str, owner: str) -> TaskClaim:
        """手动认领指定 ready 任务。"""

    def claim_next(self, owner: str) -> TaskClaim | None:
        """自动认领第一个 ready 任务。"""

    def complete_task(self, task_id: str, owner: str, claim_token: str) -> TaskCompletion:
        """使用当前租约凭证完成任务。"""


class TaskClaimService(Protocol):
    """把 Agent 身份绑定到仓库操作的应用服务接口。"""

    @property
    def store(self) -> LeasedTaskStore:
        """返回该 Service 使用的唯一仓库实例。"""

    def claim_task(self, task_id: str, context: ToolContext) -> TaskClaim:
        """从 ToolContext 获取 owner，模型参数不能伪造身份。"""

    def claim_next(self, owner: str) -> TaskClaim | None:
        """由受管 worker 使用自己的固定身份自动认领。"""

    def complete_task(
        self, task_id: str, claim_token: str, context: ToolContext
    ) -> TaskCompletion:
        """使用上下文身份和模型回传 token 完成任务。"""


class DirectTaskClaimService:
    """不保存额外状态的默认任务认领 Service。"""

    def __init__(self, store: LeasedTaskStore) -> None:
        """只保存组合根提供的共享 Repository 引用。"""
        self._store = store

    @property
    def store(self) -> LeasedTaskStore:
        """暴露引用，供组合根校验所有 Agent 是否真的共享一个仓库。"""
        return self._store

    def claim_task(self, task_id: str, context: ToolContext) -> TaskClaim:
        """把可信 ``context.identity`` 写成 owner。"""
        return self._store.claim_task(task_id, context.identity)

    def claim_next(self, owner: str) -> TaskClaim | None:
        """让队友后台线程直接以自己的固定 name 认领。"""
        return self._store.claim_next(owner)

    def complete_task(
        self, task_id: str, claim_token: str, context: ToolContext
    ) -> TaskCompletion:
        """同时校验 identity 和 claim token。"""
        return self._store.complete_task(task_id, context.identity, claim_token)


class WorkStealingSleeper(Protocol):
    """可中断等待接口；测试可注入即时 sleeper，避免真的等 5 秒。"""

    def sleep(self, seconds: float, wakeup: threading.Event) -> None:
        """等待超时或等待 wakeup 被 set。"""


class EventWorkStealingSleeper:
    """基于 ``threading.Event.wait`` 的生产 sleeper。"""

    def sleep(self, seconds: float, wakeup: threading.Event) -> None:
        """新消息或关闭流程调用 ``set()`` 后会立即结束等待。"""
        wakeup.wait(seconds)


class WorkStealingRuntime:
    """封装共享 Store、认领 Service、轮询参数和两套工具列表。

    字段说明：
        ``_store``：Lead、子 Agent、队友共同操作的 SQLite Repository。
        ``_claim_service``：统一绑定 owner 的领域 Service。
        ``_sleeper``：可被新消息和关闭流程中断的等待器。
        ``_poll_interval_seconds``：两次空扫描之间等待多久。
        ``_max_idle_polls``：连续多少次空扫描后结束当前 worker 线程。
        ``_lead_tool_definitions``：完整五工具，包含 create_task。
        ``_teammate_tool_definitions``：去掉 create_task 的四工具。
    """

    def __init__(
        self,
        store: LeasedTaskStore,
        *,
        claim_service: TaskClaimService | None = None,
        sleeper: WorkStealingSleeper | None = None,
        poll_interval_seconds: float = 5.0,
        max_idle_polls: int = 12,
    ) -> None:
        """固定本次运行所用的仓库、服务和轮询策略。"""
        _require_leased_store(store)
        actual_service = claim_service or DirectTaskClaimService(store)
        if actual_service.store is not store:
            raise ValueError("claim_service 必须使用 WorkStealingRuntime 的同一个 store")
        actual_sleeper = sleeper or EventWorkStealingSleeper()
        if not callable(getattr(actual_sleeper, "sleep", None)):
            raise TypeError("sleeper 必须实现 sleep()")
        if not isinstance(poll_interval_seconds, (int, float)) or isinstance(
            poll_interval_seconds, bool
        ):
            raise TypeError("poll_interval_seconds 必须是数字")
        if poll_interval_seconds <= 0:
            raise ValueError("poll_interval_seconds 必须大于 0")
        if isinstance(max_idle_polls, bool) or not isinstance(max_idle_polls, int):
            raise TypeError("max_idle_polls 必须是整数")
        if max_idle_polls <= 0:
            raise ValueError("max_idle_polls 必须大于 0")
        self._store = store
        self._claim_service = actual_service
        self._sleeper = actual_sleeper
        self._poll_interval_seconds = float(poll_interval_seconds)
        self._max_idle_polls = max_idle_polls
        definitions = leased_task_tool_definitions(store, actual_service)
        self._lead_tool_definitions = definitions
        self._teammate_tool_definitions = definitions[1:]

    @property
    def store(self) -> LeasedTaskStore:
        """返回共享仓库实例。"""
        return self._store

    @property
    def claim_service(self) -> TaskClaimService:
        """返回共享身份绑定 Service。"""
        return self._claim_service

    @property
    def max_idle_polls(self) -> int:
        """返回连续空轮询上限。"""
        return self._max_idle_polls

    @property
    def lead_tool_definitions(self) -> tuple[ToolDefinition, ...]:
        """Lead 使用完整五工具。"""
        return self._lead_tool_definitions

    @property
    def teammate_tool_definitions(self) -> tuple[ToolDefinition, ...]:
        """持续队友使用四工具，不能创建新任务。"""
        return self._teammate_tool_definitions

    def claim_next(self, owner: str) -> TaskClaim | None:
        """委托共享 Service 自动认领下一个 ready 任务。"""
        return self._claim_service.claim_next(owner)

    def wait_for_poll(self, wakeup: threading.Event) -> None:
        """等待下一次轮询；wakeup 被设置时提前返回。"""
        self._sleeper.sleep(self._poll_interval_seconds, wakeup)

    def render_claim_prompt(self, claim: TaskClaim) -> str:
        """把租约和任务渲染成确定性的模型输入。"""
        payload = _claim_payload(claim)
        encoded = json.dumps(payload, ensure_ascii=False, separators=(",", ":"), sort_keys=True)
        return f"<auto-claimed-task>\n{encoded}\n</auto-claimed-task>"


def leased_task_tool_definitions(
    store: LeasedTaskStore,
    claim_service: TaskClaimService | None = None,
) -> tuple[ToolDefinition, ...]:
    """创建 Lead 的五个 SQLite Task 工具定义。"""
    _require_leased_store(store)
    service = claim_service or DirectTaskClaimService(store)
    if service.store is not store:
        raise ValueError("claim_service 必须使用同一个 leased Task store")
    return (
        ToolDefinition(
            "create_task",
            "创建带显式依赖的 SQLite 项目任务。",
            _create_parameters(),
            "write",
            lambda arguments, _context: _task_operation(
                lambda: tool_success(
                    _encode_payload(
                        _task_payload(
                            store.create_task(
                                CreateTaskInput(
                                    str(arguments["subject"]),
                                    str(arguments.get("description", "")),
                                    tuple(str(item) for item in arguments.get("blocked_by", ())),
                                )
                            )
                        )
                    )
                )
            ),
            _validate_create,
        ),
        ToolDefinition(
            "get_task",
            "按 UUID 读取一个 SQLite 项目任务。",
            _task_id_parameters(),
            "read",
            lambda arguments, _context: _task_operation(
                lambda: tool_success(
                    _encode_payload(_task_payload(store.get_task(str(arguments["task_id"]))))
                )
            ),
            _validate_task_id,
        ),
        ToolDefinition(
            "list_tasks",
            "按创建顺序列出 SQLite 任务图。",
            {"type": "object", "properties": {}, "additionalProperties": False},
            "read",
            lambda _arguments, _context: _task_operation(
                lambda: tool_success(
                    _encode_payload({"tasks": [_task_payload(task) for task in store.list_tasks()]})
                )
            ),
            lambda arguments: not arguments,
        ),
        ToolDefinition(
            "claim_task",
            "原子认领 ready 任务，并返回 claim token 和租约截止时间。",
            _task_id_parameters(),
            "write",
            lambda arguments, context: _task_operation(
                lambda: tool_success(
                    _encode_payload(service.claim_task(str(arguments["task_id"]), context))
                )
            ),
            _validate_task_id,
        ),
        ToolDefinition(
            "complete_task",
            "使用当前 identity 和 claim token 完成已认领任务。",
            _complete_parameters(),
            "write",
            lambda arguments, context: _task_operation(
                lambda: _completion_result(
                    service.complete_task(
                        str(arguments["task_id"]), str(arguments["claim_token"]), context
                    )
                )
            ),
            _validate_complete,
        ),
    )


def register_leased_task_tools(
    registry: ToolRegistry,
    store: LeasedTaskStore,
    claim_service: TaskClaimService | None = None,
) -> None:
    """给 Lead 或一次性子 Agent 注册完整五工具。"""
    for definition in leased_task_tool_definitions(store, claim_service):
        registry.register(definition)


def register_teammate_leased_task_tools(
    registry: ToolRegistry,
    store: LeasedTaskStore,
    claim_service: TaskClaimService | None = None,
) -> None:
    """给持续队友注册四工具，明确跳过 ``create_task``。"""
    for definition in leased_task_tool_definitions(store, claim_service)[1:]:
        registry.register(definition)


def canonical_claim_token(value: object) -> str:
    """claim token 必须是小写 canonical UUID。"""
    try:
        return canonical_task_id(value)
    except TaskError as error:
        raise TaskClaimError("claim token 必须是小写 canonical UUID") from error


def _task_operation(operation: Any) -> ToolResult:
    """把已知 Task 领域异常转换成稳定工具错误。"""
    try:
        result = operation()
    except TaskError as error:
        return tool_error(error.code, str(error))
    if not isinstance(result, ToolResult):
        raise TypeError("Task 工具内部操作必须返回 ToolResult")
    return result


def _completion_result(completion: TaskCompletion) -> ToolResult:
    """把领域完成结果转换成模型可见 JSON。"""
    return tool_success(
        _encode_payload(
            {
                "task": _task_payload(completion.task),
                "unblocked": [_task_payload(task) for task in completion.unblocked],
            }
        )
    )


def _task_payload(task: Task) -> dict[str, object]:
    """把 Python 字段转换成稳定 wire format。"""
    return {
        "blocked_by": list(task.blocked_by),
        "description": task.description,
        "id": task.id,
        "owner": task.owner,
        "status": task.status,
        "subject": task.subject,
    }


def _claim_payload(claim: TaskClaim) -> dict[str, object]:
    """把 ``TaskClaim`` 转换成模型需要回传的三个字段。"""
    return {
        "claim_token": claim.claim_token,
        "lease_expires_at_utc": claim.lease_expires_at_utc.astimezone(UTC)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z"),
        "task": _task_payload(claim.task),
    }


def _encode_payload(value: object) -> str:
    """生成确定性紧凑 JSON；dataclass TaskClaim 会先转换为普通字典。"""
    if isinstance(value, TaskClaim):
        value = _claim_payload(value)
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def _create_parameters() -> dict[str, Any]:
    """返回 create_task 的 JSON Schema。"""
    return {
        "type": "object",
        "properties": {
            "subject": {"type": "string", "minLength": 1},
            "description": {"type": "string"},
            "blocked_by": {
                "type": "array",
                "items": {"type": "string", "pattern": CANONICAL_UUID.pattern},
                "uniqueItems": True,
            },
        },
        "required": ["subject"],
        "additionalProperties": False,
    }


def _task_id_parameters() -> dict[str, Any]:
    """返回只接收 task_id 的 JSON Schema。"""
    return {
        "type": "object",
        "properties": {"task_id": {"type": "string", "pattern": CANONICAL_UUID.pattern}},
        "required": ["task_id"],
        "additionalProperties": False,
    }


def _complete_parameters() -> dict[str, Any]:
    """返回 complete_task 的 JSON Schema，claim_token 是必填权限证明。"""
    return {
        "type": "object",
        "properties": {
            "task_id": {"type": "string", "pattern": CANONICAL_UUID.pattern},
            "claim_token": {"type": "string", "pattern": CANONICAL_UUID.pattern},
        },
        "required": ["task_id", "claim_token"],
        "additionalProperties": False,
    }


def _validate_create(arguments: Mapping[str, Any]) -> bool:
    """严格校验 create_task 输入，不接受未知字段。"""
    if not set(arguments).issubset({"subject", "description", "blocked_by"}):
        return False
    subject = arguments.get("subject")
    description = arguments.get("description", "")
    dependencies = arguments.get("blocked_by", ())
    return (
        isinstance(subject, str)
        and bool(subject.strip())
        and isinstance(description, str)
        and isinstance(dependencies, (list, tuple))
        and all(
            isinstance(item, str) and CANONICAL_UUID.fullmatch(item) is not None
            for item in dependencies
        )
        and len(dependencies) == len(set(dependencies))
    )


def _validate_task_id(arguments: Mapping[str, Any]) -> bool:
    """严格校验单 task_id 输入。"""
    return (
        set(arguments) == {"task_id"}
        and isinstance(arguments["task_id"], str)
        and CANONICAL_UUID.fullmatch(arguments["task_id"]) is not None
    )


def _validate_complete(arguments: Mapping[str, Any]) -> bool:
    """严格校验 task_id 与 claim_token。"""
    return (
        set(arguments) == {"task_id", "claim_token"}
        and all(
            isinstance(arguments[name], str)
            and CANONICAL_UUID.fullmatch(arguments[name]) is not None
            for name in ("task_id", "claim_token")
        )
    )


def _require_leased_store(value: object) -> None:
    """运行时检查最小结构契约，错误配置在启动阶段立即暴露。"""
    methods = (
        "create_task",
        "get_task",
        "list_tasks",
        "claim_task",
        "claim_next",
        "complete_task",
    )
    if not all(callable(getattr(value, name, None)) for name in methods):
        raise TypeError("store 必须实现 LeasedTaskStore")


def require_claim(value: TaskClaim | None) -> TaskClaim:
    """教学辅助：把可选认领转换成必有值，否则给出清晰异常。"""
    if value is None:
        raise TaskStorageError("当前没有可认领的 ready 任务")
    return value

File: typescript/chapters/ch20/src/features/work-stealing.ts
// 任务认领与完成能力：Lead、子代理和持续 Teammate 通过同一 SQLite store 与 claim service 操作任务，避免各自读写独立状态。
import { z } from "zod";

import type { ToolContext, ToolDefinition, ToolRegistry, ToolResult } from "../core/tools.js";
import { toolError, toolSuccess } from "../core/tools.js";
import { canonicalTaskId, type Task, TaskError } from "./tasks.js";
import type { CreateTaskInput, TaskCompletion } from "./tasks.js";

const uuidSchema = z
  .string()
  .regex(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/u);
const createTaskSchema = z.strictObject({
  subject: z.string().trim().min(1),
  description: z.string().trim().default(""),
  blocked_by: z.array(uuidSchema).default([]),
});
const taskIdSchema = z.strictObject({ task_id: uuidSchema });
const listTasksSchema = z.strictObject({});
const completeTaskSchema = z.strictObject({ task_id: uuidSchema, claim_token: uuidSchema });

export class TaskClaimError extends TaskError {
  // 表示调用方提交的 owner/token 与当前租约不匹配，可稳定映射为工具错误码。
  constructor(message: string, code = "task_claim_mismatch") {
    super(code, message);
    this.name = "TaskClaimError";
  }
}

export class TaskLeaseExpiredError extends TaskClaimError {
  // 单独区分“曾经有效但已过期”，便于模型重新认领而不是重复完成。
  constructor(message: string) {
    super(message, "task_lease_expired");
    this.name = "TaskLeaseExpiredError";
  }
}

export interface TaskClaim {
  // claimToken 是完成权限证明，owner 相同也不能用旧租约完成被重新认领的任务。
  readonly task: Task;
  readonly claimToken: string;
  readonly leaseExpiresAtUtc: Date;
}

export interface LeasedTaskStore {
  // 存储层负责 DAG、租约与 token 的原子语义，工具层不直接拼接持久化操作。
  createTask(input: CreateTaskInput): Promise<Task>;
  getTask(taskId: string): Promise<Task>;
  listTasks(): Promise<readonly Task[]>;
  claimTask(taskId: string, owner: string): Promise<TaskClaim>;
  claimNext(owner: string): Promise<TaskClaim | undefined>;
  completeTask(taskId: string, owner: string, claimToken: string): Promise<TaskCompletion>;
}

export interface TaskClaimService {
  // service 将执行身份绑定到存储 owner，阻止模型通过工具参数伪造其他 Agent。
  readonly store: LeasedTaskStore;
  claimTask(taskId: string, context: ToolContext): Promise<TaskClaim>;
  claimNext(owner: string): Promise<TaskClaim | undefined>;
  completeTask(taskId: string, claimToken: string, context: ToolContext): Promise<TaskCompletion>;
}

// 默认服务把 ToolContext.identity 映射为 owner，不让工具输入伪造认领者。
class DirectTaskClaimService implements TaskClaimService {
  readonly #store: LeasedTaskStore;

  constructor(store: LeasedTaskStore) {
    // service 不复制状态，只持有组合根提供的唯一 store 引用。
    this.#store = store;
  }

  get store(): LeasedTaskStore {
    // 暴露引用身份供 WorkStealingRuntime 拒绝错误配对的 service。
    return this.#store;
  }

  async claimTask(taskId: string, context: ToolContext): Promise<TaskClaim> {
    // 手动工具认领的 owner 固定取自 ToolContext.identity。
    return await this.#store.claimTask(taskId, context.identity);
  }

  async claimNext(owner: string): Promise<TaskClaim | undefined> {
    // 自动扫描不是模型工具调用，owner 由调用方直接传入；它不经过 ToolContext，模型也不能伪造。
    return await this.#store.claimNext(owner);
  }

  async completeTask(
    taskId: string,
    claimToken: string,
    context: ToolContext,
  ): Promise<TaskCompletion> {
    // 完成同样从上下文绑定 owner，并要求模型回传当前 claim token。
    return await this.#store.completeTask(taskId, context.identity, claimToken);
  }
}

export interface WorkStealingSleeper {
  // wakeup 可由新消息、协议事件或关闭流程中断，避免 worker 必须等满轮询周期。
  sleep(seconds: number, wakeup: AbortSignal): Promise<void>;
}

// Lead 使用完整五工具元组；Teammate 通过切片移除 create_task。
type LeasedToolDefinitions = ReturnType<typeof leasedTaskToolDefinitions>;
type TeammateLeasedToolDefinitions = readonly LeasedToolDefinitions[number][];

// AsyncioWorkStealingSleeper 默认 sleeper 使用 timer + abort 监听，避免空闲轮询无法被关闭流程中断。
export class AsyncioWorkStealingSleeper implements WorkStealingSleeper {
  async sleep(seconds: number, wakeup: AbortSignal): Promise<void> {
    // 已触发的 wakeup 直接返回，避免遗漏绑定监听前发生的唤醒。
    if (wakeup.aborted) return;
    await new Promise<void>((resolve) => {
      // 同时监听 timer 和 abort：新消息到达时提前结束等待，避免 idle worker 卡满 5 秒。
      const timer = setTimeout(finish, seconds * 1000);
      const onAbort = (): void => {
        clearTimeout(timer);
        finish();
      };
      function finish(): void {
        wakeup.removeEventListener("abort", onAbort);
        resolve();
      }
      wakeup.addEventListener("abort", onAbort, { once: true });
      if (wakeup.aborted) onAbort();
    });
  }
}

// WorkStealingRuntime 封装任务认领/完成、轮询参数与工具定义，让 bootstrap 只关注能力组合而无需介入认领细节。
export class WorkStealingRuntime {
  // 运行时只通过共享 TaskClaimService 认领与完成，避免多队友直接竞争存储细节。
  readonly #store: LeasedTaskStore;
  readonly #claimService: TaskClaimService;
  readonly #sleeper: WorkStealingSleeper;
  readonly #pollIntervalSeconds: number;
  readonly #maxIdlePolls: number;
  readonly #leadToolDefinitions: LeasedToolDefinitions;
  readonly #teammateToolDefinitions: TeammateLeasedToolDefinitions;

  constructor(options: {
    readonly store: LeasedTaskStore;
    readonly claimService?: TaskClaimService;
    readonly sleeper?: WorkStealingSleeper;
    readonly pollIntervalSeconds?: number;
    readonly maxIdlePolls?: number;
  }) {
    // 构造时冻结 store/service/轮询策略和工具快照，运行中不允许热切换认领语义。
    if (!isLeasedTaskStore(options.store))
      throw new TypeError("store must implement LeasedTaskStore");
    const claimService =
      options.claimService === undefined
        ? new DirectTaskClaimService(options.store)
        : options.claimService;
    // claimService 必须指向同一个 store，否则 Lead 与 Teammate 可能操作两个数据库。
    if (!isTaskClaimService(claimService))
      throw new TypeError("claimService must implement TaskClaimService");
    if (claimService.store !== options.store) {
      throw new Error("claimService must use the WorkStealingRuntime store");
    }
    const sleeper =
      options.sleeper === undefined ? new AsyncioWorkStealingSleeper() : options.sleeper;
    if (typeof sleeper.sleep !== "function") throw new TypeError("sleeper must implement sleep()");
    const pollIntervalSeconds =
      options.pollIntervalSeconds === undefined ? 5 : options.pollIntervalSeconds;
    const maxIdlePolls = options.maxIdlePolls === undefined ? 12 : options.maxIdlePolls;
    if (!Number.isFinite(pollIntervalSeconds) || pollIntervalSeconds <= 0) {
      throw new RangeError("pollIntervalSeconds must be positive");
    }
    if (!Number.isInteger(maxIdlePolls) || maxIdlePolls <= 0) {
      throw new RangeError("maxIdlePolls must be a positive integer");
    }
    this.#store = options.store;
    this.#claimService = claimService;
    this.#sleeper = sleeper;
    this.#pollIntervalSeconds = pollIntervalSeconds;
    this.#maxIdlePolls = maxIdlePolls;
    // 同一份五工具先整体给 Lead；Teammate 从索引 1 开始，只暴露读取、认领和完成。
    const definitions = leasedTaskToolDefinitions(this.#store, this.#claimService);
    this.#leadToolDefinitions = definitions;
    this.#teammateToolDefinitions = Object.freeze(definitions.slice(1));
  }

  get store(): LeasedTaskStore {
    // 暴露同一实例供 bootstrap 校验 Lead、Subagent 与 Teammate 没有分叉存储。
    return this.#store;
  }

  get claimService(): TaskClaimService {
    // 工具注册和自动认领必须复用同一身份绑定服务。
    return this.#claimService;
  }

  get maxIdlePolls(): number {
    // TeammateRuntime 用该上限决定连续空轮询何时结束 worker。
    return this.#maxIdlePolls;
  }

  get leadToolDefinitions(): LeasedToolDefinitions {
    // Lead 可创建、查询、认领和完成任务。
    return this.#leadToolDefinitions;
  }

  get teammateToolDefinitions(): TeammateLeasedToolDefinitions {
    // 持续队友只能处理既有任务，不能在空闲循环中扩张任务图。
    return this.#teammateToolDefinitions;
  }

  async claimNext(owner: string): Promise<TaskClaim | undefined> {
    // 自动认领 owner 来自 worker 身份，不经过模型工具输入。
    return await this.#claimService.claimNext(owner);
  }

  async waitForPoll(wakeup: AbortSignal): Promise<void> {
    // 轮询等待由 sleeper 抽象，测试可用确定性实现而无需真实等待。
    await this.#sleeper.sleep(this.#pollIntervalSeconds, wakeup);
  }

  renderClaimPrompt(claim: TaskClaim): string {
    // 用确定性 JSON 把 token、lease 和任务状态一起交给模型，模型必须显式调用 complete_task。
    const payload = {
      claim_token: claim.claimToken,
      lease_expires_at_utc: claim.leaseExpiresAtUtc.toISOString(),
      task: taskPayload(claim.task),
    };
    return `<auto-claimed-task>\n${JSON.stringify(payload)}\n</auto-claimed-task>`;
  }
}

export function leasedTaskToolDefinitions(
  store: LeasedTaskStore,
  claimService: TaskClaimService = new DirectTaskClaimService(store),
): readonly [
  ToolDefinition<z.infer<typeof createTaskSchema>>,
  ToolDefinition<z.infer<typeof taskIdSchema>>,
  ToolDefinition<z.infer<typeof listTasksSchema>>,
  ToolDefinition<z.infer<typeof taskIdSchema>>,
  ToolDefinition<z.infer<typeof completeTaskSchema>>,
] {
  if (!isLeasedTaskStore(store)) throw new TypeError("store must implement LeasedTaskStore");
  if (!isTaskClaimService(claimService) || claimService.store !== store) {
    throw new Error("claimService must use the leased Task store");
  }
  // 数组顺序即 Lead/Teammate 的工具分界：Teammate 注册时跳过 create_task。
  const create: ToolDefinition<z.infer<typeof createTaskSchema>> = {
    name: "create_task",
    description: "Create a persistent SQLite project task with explicit dependencies.",
    inputSchema: createTaskSchema,
    effect: "write",
    handler: async (input) => {
      try {
        return toolSuccess(
          JSON.stringify(
            taskPayload(
              await store.createTask({
                subject: input.subject,
                description: input.description,
                blockedBy: input.blocked_by,
              }),
            ),
          ),
        );
      } catch (error) {
        return leasedTaskError(error);
      }
    },
  };
  const get: ToolDefinition<z.infer<typeof taskIdSchema>> = {
    name: "get_task",
    description: "Read one persistent SQLite project task by ID.",
    inputSchema: taskIdSchema,
    effect: "read",
    handler: async (input) => {
      try {
        return toolSuccess(JSON.stringify(taskPayload(await store.getTask(input.task_id))));
      } catch (error) {
        return leasedTaskError(error);
      }
    },
  };
  const list: ToolDefinition<z.infer<typeof listTasksSchema>> = {
    name: "list_tasks",
    description: "List the SQLite project task graph in stable creation order.",
    inputSchema: listTasksSchema,
    effect: "read",
    handler: async () => {
      try {
        const tasks = await store.listTasks();
        return toolSuccess(JSON.stringify({ tasks: tasks.map(taskPayload) }));
      } catch (error) {
        return leasedTaskError(error);
      }
    },
  };
  const claim: ToolDefinition<z.infer<typeof taskIdSchema>> = {
    name: "claim_task",
    description: "Atomically claim a ready task and return its claim token and lease.",
    inputSchema: taskIdSchema,
    effect: "write",
    handler: async (input, context) => {
      try {
        // claimService 从 ToolContext.identity 推导 owner，工具参数没有 owner 字段。
        return toolSuccess(
          JSON.stringify(claimPayload(await claimService.claimTask(input.task_id, context))),
        );
      } catch (error) {
        return leasedTaskError(error);
      }
    },
  };
  const complete: ToolDefinition<z.infer<typeof completeTaskSchema>> = {
    name: "complete_task",
    description: "Complete an owned task with its current claim token.",
    inputSchema: completeTaskSchema,
    effect: "write",
    handler: async (input, context) => {
      try {
        // owner 仍由 context.identity 注入，claim_token 是模型唯一需要回传的权限证明。
        const result = await claimService.completeTask(input.task_id, input.claim_token, context);
        return toolSuccess(
          JSON.stringify({
            task: taskPayload(result.task),
            unblocked: result.unblocked.map(taskPayload),
          }),
        );
      } catch (error) {
        return leasedTaskError(error);
      }
    },
  };
  return [create, get, list, claim, complete];
}

export function registerLeasedTaskTools(
  registry: ToolRegistry,
  store: LeasedTaskStore,
  claimService?: TaskClaimService,
): void {
  // Lead 和一次性子代理可创建任务、查询、认领并完成整个 DAG。
  const [create, get, list, claim, complete] = leasedTaskToolDefinitions(store, claimService);
  registry.register(create);
  registry.register(get);
  registry.register(list);
  registry.register(claim);
  registry.register(complete);
}

export function registerTeammateLeasedTaskTools(
  registry: ToolRegistry,
  store: LeasedTaskStore,
  claimService?: TaskClaimService,
): void {
  // 持续队友只处理已有任务，不暴露 create_task，避免空闲循环无界扩张任务图。
  const [, get, list, claim, complete] = leasedTaskToolDefinitions(store, claimService);
  registry.register(get);
  registry.register(list);
  registry.register(claim);
  registry.register(complete);
}

export function canonicalClaimToken(value: string): string {
  // token 使用 canonical UUID，拒绝大小写或格式变体形成多个文本身份。
  try {
    return canonicalTaskId(value);
  } catch {
    throw new TaskClaimError("Claim token must be a canonical UUID");
  }
}

function isLeasedTaskStore(value: unknown): value is LeasedTaskStore {
  // 组合根只依赖最小结构契约，避免把具体 SQLite 类泄漏到领域层。
  return (
    typeof value === "object" &&
    value !== null &&
    typeof Reflect.get(value, "createTask") === "function" &&
    typeof Reflect.get(value, "getTask") === "function" &&
    typeof Reflect.get(value, "listTasks") === "function" &&
    typeof Reflect.get(value, "claimTask") === "function" &&
    typeof Reflect.get(value, "claimNext") === "function" &&
    typeof Reflect.get(value, "completeTask") === "function"
  );
}

function isTaskClaimService(value: unknown): value is TaskClaimService {
  // 除方法形状外还校验 store，随后构造器继续要求实例引用完全相同。
  return (
    typeof value === "object" &&
    value !== null &&
    isLeasedTaskStore(Reflect.get(value, "store")) &&
    typeof Reflect.get(value, "claimTask") === "function" &&
    typeof Reflect.get(value, "claimNext") === "function" &&
    typeof Reflect.get(value, "completeTask") === "function"
  );
}

function taskPayload(task: Task): Readonly<Record<string, unknown>> {
  // 内存 Task 使用 blockedBy，工具结果统一映射为 wire format 的 blocked_by。
  return {
    blocked_by: [...task.blockedBy],
    description: task.description,
    id: task.id,
    owner: task.owner,
    status: task.status,
    subject: task.subject,
  };
}

function claimPayload(claim: TaskClaim): Readonly<Record<string, unknown>> {
  // TaskClaim 内部字段是 camelCase，序列化给模型时统一换成 claim_token 与 lease_expires_at_utc。
  return {
    claim_token: claim.claimToken,
    lease_expires_at_utc: claim.leaseExpiresAtUtc.toISOString(),
    task: taskPayload(claim.task),
  };
}

function leasedTaskError(error: unknown): ToolResult {
  if (error instanceof TaskError) return toolError(error.code, error.message);
  throw error;
}
